lib/graphql_includable.rb

```diff
@@ -25,96 +27,128 @@ def delegate(*methods, args)
-        else
-          includes << wrap_delegate(specified_includes || association_name, delegated_model_name)
-        end
-      elsif specified_includes
-        includes << specified_includes
+    return [] unless node && return_model
+
+    node.scoped_children[node_return_type(node)].each do |_child_name, child_node|
+      child_includes = includes_from_graphql_child(child_node, return_model)
+      if child_includes.is_a?(Hash)
+        nested_includes.merge!(child_includes)
+      else
+        child_includes = [child_includes] unless child_includes.is_a?(Array)
+        includes += child_includes
       end
     end
 
-    includes << nested_includes if nested_includes.size > 0
+    includes << nested_includes unless nested_includes.empty?
     includes
   end
 
-  def self.node_has_active_record_children(node)
-    node.scoped_children[unwrapped_type(node)].each do |child_return_name, child_node|
-      node_returns_active_record?(child_node)
+  def self.includes_from_graphql_child(child_node, return_model)
+    specified_includes = child_node.definitions[0].metadata[:includes]
+    attribute_name = node_predicted_association_name(child_node)
+    includes_chain = delegated_includes_chain(return_model, attribute_name)
+
+    if model_has_association?(return_model, attribute_name, includes_chain)
+      child_includes = includes_from_graphql_field(child_node)
+      includes_chain << (specified_includes || attribute_name)
+      includes_chain << child_includes unless child_includes.empty?
+      array_to_nested_hash(includes_chain)
+    else
+      includes = []
+      includes << array_to_nested_hash(includes_chain) unless includes_chain.empty?
+      includes << specified_includes if specified_includes
+      includes
     end
   end
 
   def self.node_return_class(node)
+    # rubocop:disable Lint/HandleExceptions, Style/RedundantBegin
     begin
-      Object.const_get(unwrapped_type(node).name)
+      Object.const_get(node_return_type(node).name)
     rescue NameError
     end
+    # rubocop:enable Lint/HandleExceptions, Style/RedundantBegin
   end
 
   def self.node_returns_active_record?(node)
     klass = node_return_class(node)
     klass && klass < ActiveRecord::Base
   end
 
-  # return raw contents, or contents wrapped in a hash (for delegated associations)
-  def self.wrap_delegate(contents, delegate, delegate_key = delegate)
-    return contents unless delegate
-
-    obj = {}
-    obj[delegate_key] = contents
-    obj
+  def self.node_predicted_association_name(node)
+    definition = node.definitions[0]
+    specified_includes = definition.metadata[:includes]
+    if specified_includes.is_a?(Symbol)
+      specified_includes
+    else
+      (definition.property || definition.name).to_sym
+    end
   end
 
   # unwrap GraphQL ListType and NonNullType wrappers
-  def self.unwrapped_type(node)
+  def self.node_return_type(node)
     type = node.return_type
     type = type.of_type while type.respond_to? :of_type
```

Comment:
probably makes sense to move the comment above the method to here, and expand on the comment as to *why* it is unwrapping